src/pybind/mgr/mgr_util.py

```diff
@@ -176,3 +178,124 @@ def verify_tls_files(cert_fname, pkey_fname):
         logger.warning(
             'Private key {} and certificate {} do not match up: {}'.format(
                 pkey_fname, cert_fname, str(e)))
+
+def get_most_recent_rate(rates):
+    """ Get most recent rate from rates
+
+    :param rates: The derivative between all time series data points
+    :type rates: list[tuple[float, float]]
+
+    :return: The last derivative or 0.0 if none exists
+    :rtype: float
+
+    >>> get_most_recent_rate([])
+    0.0
+    >>> get_most_recent_rate([(1, -2.0)])
+    -2.0
+    >>> get_most_recent_rate([(1, 2.0), (2, 1.5), (3, 5.0)])
+    5.0
+    """
+    if len(rates) == 0:
+        return 0.0
+    return rates[-1][1]
+
+def get_time_series_rates(data, use_abs=None):
+    """ Rates from time series data
+
+    :param data: Time series data where the time is provided in seconds
+    :type data: list[tuple[float, float]]
+
+    :param use_abs: If the derivative returned should be an absolute one
+    :type use_abs: bool (default None)
+        Using the absolute value is a temporary workaround for the pg decrease issue, that will
+        reduce the maximum amount in all time series data of a pool. Normally the maximum amount
+        should always increase.
+        https://tracker.ceph.com/issues/41795
+
+    :return: The derivative between all time series data points
+    :rtype: list[tuple[float, float]]
+
+    >>> get_time_series_rates([])
+    []
+    >>> get_time_series_rates([[1, 42]])
+    []
+    >>> get_time_series_rates([[0, 2], [0, 2]])
+    []
+    >>> get_time_series_rates([[0, 0], [1, -2]])
+    [(1, -2.0)]
+    >>> get_time_series_rates([[0, 0], [0, 0.5], [2, 1], [3, 0], [4, -2]])
+    [(2, 0.25), (3, -1.0), (4, -2.0)]
+    >>> get_time_series_rates([[1, 0], [3, 1], [4, 0], [5, -2]], True)
+    [(3, 0.5), (4, 1.0), (5, 2.0)]
+    """
+    data = _filter_time_series(data)
+    if len(data) == 0:
+        return []
+    return [(data2[0], _derivative(data1, data2, use_abs)) for data1, data2 in
+            pairwise(data)]
+
+def _filter_time_series(data):
+    """ Filters invalid samples from time series data
+    It also enforces the list to contain at least two samples.
+
+    :param data: Time series data where the time is provided in seconds
+    :type data: list[tuple[float, float]]
+
+    :return: Filtered time series data
+    :rtype: list[tuple[float, float]]
+
+    >>> _filter_time_series([])
+    []
+    >>> _filter_time_series([[1, 42]])
+    []
+    >>> _filter_time_series([[0, 2], [0, 2]])
+    []
+    >>> _filter_time_series([[0, 1], [1, 2]])
+    [[0, 1], [1, 2]]
+    >>> _filter_time_series([[0, 2], [0, 2], [0, 1], [1, 2], [1, 3]])
+    [[0, 1], [1, 3]]
+    >>> _filter_time_series([[0, 0], [0, 0.5], [2, 1], [3, 0], [4, -2]])
+    [[0, 0.5], [2, 1], [3, 0], [4, -2]]
+    """
+    if data is None:
+        return []
+    max = len(data)-1
+    data = [d for i, d in enumerate(data) if i < max and d[0] != data[i+1][0] or i == max]
```

Comment:
Yes it was a bit hard to read that's why I created a lambda which now is used inside the loop. This way it's easier to revert the outcome in case of a sample with the same timestamp as another.

Hope my solution suits you as well.